rfc/rfc-73/rfc-73.md

```diff
@@ -0,0 +1,416 @@
+   single table.
+8. **Configurability:** Allow users to configure the behavior of multi-table transactions, e.g., concurrency control
+   mechanism, timeout durations, etc.
+9. **Recovery**: Rollbacks, savepoint and restore should be supported.
+
+### Non-goals
+
+1. **Cross-Database Transactions:** Transactions spanning multiple databases might be too complex as an initial target.
+2. **Granular Record Locking:** Instead of locking at the granularity of individual records, coarse-grained locking (
+   like at the file level) might be more practical to start with. But, let's avoid table or partition level locking as
+   mentioned in the goals.
+3. **Distributed Transactions:** We are not going to consider transactions in the sense of distributed databases, and
+   thus the need of a protocol such as 2PC or a quorum mechanism to proceed with the commits.
+4. **Complex Conflict Resolution:** Initially, simple conflict resolution strategies (like aborting a transaction on
+   conflict) can be implemented, leaving more sophisticated strategies for future iterations.
+5. **Replication/CDC stream:** We are not going to build a replication or CDC stream out of the database level commit
+   log, though that would be a good usage of the database timeline introduced in this RFC.
+
+### Design
+
+Our primary challenge is to support operations that span multiple tables within a single database, and maintain the ACID
+properties across these tables.
+
+1. Need for a catalog: Do we need a catalog API that tracks databases and its tables?
+2. Need for a transaction coordinator: A centralized coordinator that will manage transaction logs, track ongoing
+   transactions, handle timeouts, and manage transaction rollbacks.
+3. Need for a transaction log: At the table level, the timeline incorporating state of an action with start and modified
+   time serves the purpose of transaction log. At the database level, we need to track all multi-table transactions.
+   Every start, update, or commit of a multi-table transaction gets recorded (Database timeline?).
+4. Locking/Conflict resolution mechanism: Lock the affected files during the multi-table transaction to prevent
+   conflicting writes. Decide on conflict resolution strategies (e.g., last write wins, version vectors).
+5. Need for buffer management: Writes by a transaction are not immediately visible to other transactions. They are
+   buffered until the transaction decides to commit. In-memory vs durable buffer (reuse `.temp` dir)?
+
+#### Concurrency Control and Conflict Resolution
+
+Today we have an event log at the table level in the form of Hudi timeline. To support multi-table transactions, we are
+going to need a unified view of all tables within the database to ensure atomicity and consistency. Hence, we propose a
+database-level timeline as transaction log, which mainly contains the following:
+
+* Transaction ID
+* Tables involved
+* Type of operation (e.g., write, delete)
```

Comment:
again, txn can be a mix of these?

Reply:
good point. yes it can be a mix of these. let me think a bit more.